### PR TITLE
New node: Math

### LINKED ISSUE #2115: Math node for running custom expressions with A and B
The new maths parser introduced in #2033 is currently only utilised to immediately evaluate expressions in number inputs.

To expand its utility, a new node should be introduced to run an expression during the runtime of the graph. In future, this node could take inputs from the graph. However this is blocked on implementing nodes with variadic inputs.

### PR DESCRIPTION
**Pull request by jacinyan**
Closes #2115

Add runtime Expression node

This PR introduces an Expression node that evaluates mathematical expressions during graph runtime. This is an initial implementation that currently only supports standalone expressions. Support for graph inputs is blocked pending variadic inputs implementation.

Changes:
- Implement basic ExpressionNode that evaluates math expressions using math_parser
- Register the Expression node in the node registry
- Add DocumentNode Definition for Expression node
- Add type system support in network_interface for Expression nodes

Current limitations:
- Can only evaluate expressions without inputs from the graph
- Will be expanded to support graph inputs once variadic inputs are implemented

Testing:
- Unit test added for basic expression evaluation

Please advise correction and improvement.

### CODE REVIEW

File: node-graph/graph-craft/src/expression.rs
Comment (0HyperCube) [inline node-graph/graph-craft/src/expression.rs:16]: Nodes should not return tagged value since then the output cannot be used by any other nodes.

File: node-graph/graph-craft/src/expression.rs
Comment (0HyperCube) [inline node-graph/graph-craft/src/expression.rs:12]: Please consider using the node macro which simplifies node implementations.

File: editor/src/messages/portfolio/document/utility_types/network_interface.rs
Comment (0HyperCube) [inline editor/src/messages/portfolio/document/utility_types/network_interface.rs:526]: I'm not really sure why you have added an additional node implementation.

File: editor/src/messages/portfolio/document/node_graph/document_node_definitions.rs
Comment (0HyperCube) [inline editor/src/messages/portfolio/document/node_graph/document_node_definitions.rs:2537]: It would be nice if the properties allowed the user to input the string for the expression.

File: editor/src/messages/portfolio/document/node_graph/document_node_definitions.rs
Comment (0HyperCube) [inline editor/src/messages/portfolio/document/node_graph/document_node_definitions.rs:2521]: Using the node macro, this can be automatically generated.

File: node-graph/graph-craft/src/lib.rs
Comment (0HyperCube) [inline node-graph/graph-craft/src/lib.rs:19]: Could this perhaps be moved to the graphene-core?

**Review by 0HyperCube — COMMENTED** [review]
Thanks for your work on this so far.

The goal of the expression node is to allow the node's output to be used within the graph (e.g. to set the width of the rectangle) as seen below. If you have any questions regarding this, feel free to ask in the discord chat.
image

Thanks again for this contribution.

**Comment by jacinyan** [comment]
> Thanks for your work on this so far.
> 
> The goal of the expression node is to allow the node's output to be used within the graph (e.g. to set the width of the rectangle) as seen below. If you have any questions regarding this, feel free to ask in the discord chat. image
> 
> Thanks again for this contribution.

Thanks for the feedback @0HyperCube . I jumped in a bit too quickly without fully understanding the context. Apologies for the confusion! Could you let me know which Discord channel is best to ask questions so I can clarify what I should be doing next? Thanks again for your patience and help.

**Comment by 0HyperCube** [comment]
@jacinyan No problem. You can ask in #task-help or #development.

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
!build

**Comment by 0HyperCube** [comment]
!build

**Review by 0HyperCube — APPROVED** [review]
This works well, thanks for your work on this.

I changed the default of the input to be not exposed and set to "1 + 1", which should make it clearer about how the node works.

**Comment by jacinyan** [comment]
> This works well, thanks for your work on this.
> 
> I changed the default of the input to be not exposed and set to "1 + 1", which should make it clearer about how the node works.

Hi @0HyperCube, thanks for reviewing and approving the changes! I've read GitHub's documentation about workflows approval for first-time contributors, and I believe the remaining steps are on the maintainers' side. Just wanted to confirm if my understanding is correct?

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
Ok, with some further changes, this finally meets my original goals for the Math node.

**Review by Keavon — APPROVED** [review]

**Comment by Keavon** [comment]
Deferred for later:
- Making this work with DVec2
- Deleting all the other math nodes (with an upgrade script on document import to convert into the equivalent Math nodes), blocked by DVec2 support
- Visual error handling when an expression is invalid
- Fixing some issues in the expression calculator library for things like `abs()` and other math functions

**Pull request merged**
